### PR TITLE
server: alloc timestamp after recover

### PR DESCRIPTION
**Pull request by rleungx**
## What have you changed? (required)
This PR allows allocating timestamp after recovering from the incorrect system time.

## What are the type of the changes (required)?
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (required)?
- Manual tests

## Refer to a related PR or issue link (optional)
#499, #940, pingcap/tidb#5709.

### CODE REVIEW

File: server/tso.go [hunk 1]
```diff
@@ -82,17 +82,14 @@ func (s *Server) syncTimestamp() error {
...
-		}
-		break
+	now := time.Now()
+	// If the current system time is less than the saved etcd timestamp,
+	// which means that the system time may be incorrect,
+	// the timestamp allocation will start from the saved etcd timestamp temporarily.
+	if subTimeByWallClock(last, now)+updateTimestampGuard > 0 {
+		log.Errorf("system time may be incorrect, last: %v, now: %v", last, now)
+		s.ts.Store(&atomicObject{physical: last})
+		return nil
```
Comment (disksing) [inline server/tso.go]: I think you want to update `now` here. Or the incorrect system time will be saved.

File: server/tso.go [hunk 2]
```diff
@@ -163,10 +184,21 @@ func (s *Server) getRespTS(count uint32) (pdpb.Timestamp, error) {
...
 		resp.Physical = current.physical.UnixNano() / int64(time.Millisecond)
 		resp.Logical = atomic.AddInt64(&current.logical, int64(count))
+
 		if resp.Logical >= maxLogical {
-			log.Errorf("logical part outside of max logical interval %v, please check ntp time, retry count %d", resp, i)
-			time.Sleep(updateTimestampStep)
-			continue
+			// If the time window is enough, it will increase the physical time,
+			// otherwise, it will wait for updating the time window.
+			if subTimeByWallClock(s.lastSavedTime, current.physical) > updateTimestampGuard {
```
Comment (disksing) [inline server/tso.go]: I think `lastSavedTime` is accessed concurrently without any mutex.

Comment (rleungx) [reply server/tso.go]: agree.

File: server/tso.go [hunk 3]
```diff
@@ -122,29 +123,50 @@ func (s *Server) updateTimestamp() error {
...
+			tsoCounter.WithLabelValues("skip_save").Inc()
+			log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
+			return nil
 		}
-
-		tsoCounter.WithLabelValues("save_ok").Inc()
-		log.Debugf("save timestamp ok: prev %v last %v save %v", prev, last, save)
+		// If the time window is no more than updateTimestampGuard,
+		// it will adjust the time window.
+		if subTimeByWallClock(saved, now) <= updateTimestampGuard {
```
Comment (nolouch) [inline server/tso.go]: better also validate the `TsoSaveInterval` large than `updateTimestampGuard` in config.

File: server/tso.go [hunk 4]
```diff
@@ -122,29 +123,50 @@ func (s *Server) updateTimestamp() error {
...
+			if err := s.saveTimestamp(save); err != nil {
+				return errors.Trace(err)
+			}
+
+			tsoCounter.WithLabelValues("save_incorrect").Inc()
+			log.Debugf("save timestamp incorrect: now %v prev %v last %v save %v", now, prev, last, save)
+		}
+		// Since the physical timestamp is greater than the current system time,
+		// the physical timestamp will be used to alloc timestamp temporarily.
+		current.physical = prev
```
Comment (nolouch) [inline server/tso.go]: The logical part will clean to 0, but you do not increase the physical part.

**Comment by nolouch** [comment]
PTAL @liukun4515

File: server/tso.go (hunk 3 above)
Comment (rleungx) [reply server/tso.go]: They are not in the same order of magnitude, and thus it doesn't make much sense. What do you think?

File: server/tso.go [hunk 5]
```diff
@@ -163,10 +185,22 @@ func (s *Server) getRespTS(count uint32) (pdpb.Timestamp, error) {
...
-			continue
+			// If the time window is enough, it will increase the physical time,
+			// otherwise, it will wait for updating the time window.
+			saved := s.lastSavedTime.Load().(time.Time)
+			if subTimeByWallClock(saved, current.physical) > updateTimestampGuard {
+				last := &atomicObject{
+					physical: current.physical.Add(time.Millisecond),
+					logical:  0,
+				}
+				s.ts.Store(last)
```
Comment (disksing) [inline server/tso.go]: miss a `continue`?

**Comment by disksing** [comment]
I think we can refactor `updateTimestamp()` like this:
```
func updateTimestamp() {
    var next int64                             // step1. determine next physical timestamp
    if now > prev.physical + 1ms {            // a. if system is greater, sync with system time
        next = now
    }
    if prev.logical > maxLogical/2 {          // b. if logical is about to use up, increase physical by 1ms
        next = prev.physical+1ms
    }
    if next == 0 {                                      // otherwise the physical does not need update
        return
    }

    if saved - next <= 1ms {                  // step2. determine if we need to save ts to etcd
        saveTimestamp(next + 3s)          // when it is unsafe to increase physical to `next`, we save etcd so the saved will increase by 3s
    }

    s.ts.Store({next, 0})        // step3. update physical
    return nil
}
```

File: server/tso.go [hunk 6]
```diff
@@ -82,69 +82,65 @@ func (s *Server) syncTimestamp() error {
...
-	if since <= updateTimestampGuard {
-		tsoCounter.WithLabelValues("skip_save").Inc()
-		log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
+
+	var next time.Time
+	if jetLag > updateTimestampGuard {
+		next = now
+	} else if atomic.LoadInt64(&prev.logical) > maxLogical/2 {
+		next = prev.physical.Add(time.Millisecond)
+	} else {
```
Comment (disksing) [inline server/tso.go]: `tsoCounter.WithLabelValues("skip_save").Inc()` before return nil

File: server/tso.go [hunk 7]
```diff
@@ -82,69 +82,65 @@ func (s *Server) syncTimestamp() error {
...
-	// Avoid the same physical time stamp
-	if since <= updateTimestampGuard {
-		tsoCounter.WithLabelValues("skip_save").Inc()
-		log.Warnf("invalid physical timestamp, prev: %v, now: %v, re-update later", prev, now)
+
+	var next time.Time
+	if jetLag > updateTimestampGuard {
+		next = now
+	} else if atomic.LoadInt64(&prev.logical) > maxLogical/2 {
+		next = prev.physical.Add(time.Millisecond)
```
Comment (disksing) [inline server/tso.go]: Add warning log here.

File: server/tso.go [hunk 8]
```diff
@@ -82,69 +82,65 @@ func (s *Server) syncTimestamp() error {
...
-			time.Sleep(wait)
-			continue
-		}
-		break
+	next := time.Now()
+	// If the current system time minus the saved etcd timestamp is less than updateTimestampGuard,
+	// the timestamp allocation will start from the saved etcd timestamp temporarily.
+	if subTimeByWallClock(next, last) < updateTimestampGuard {
+		log.Errorf("system time may be incorrect: last: %v next %v", last, next)
+		next = next.Add(updateTimestampGuard)
```
Comment (disksing) [inline server/tso.go]: `next = last.Add(updateTimestampGuard)`?

**Comment by disksing** [comment]
Please review this PR carefully. /cc @nolouch @siddontang @liukun4515

**Comment by rleungx** [comment]
@disksing PTAL.

File: server/tso.go [hunk 9]
```diff
@@ -82,69 +82,83 @@ func (s *Server) syncTimestamp() error {
...
 
+// This function will do two things:
+// 1. When the logical time is going to be used up, the current physical time needs to increase.
+// 2. If the time window is not enough, which means the saved etcd time minus the next physical time
+//    is less than or equal to `updateTimestampGuard`, it will need to be updated and save the
+//    next physical time plus `TsoSaveInterval` into etcd.
+//
+// Here is some constraints that this function must satisfy:
+// 1. The physical time is monotonically increasing.
+// 2. The saved time is monotonically increasing.
```
Comment (disksing) [inline server/tso.go]: And the physical should always less than saved timestamp.

**Comment by siddontang** [comment]
please construct a test for the time fallback.

**Comment by siddontang** [comment]
why add DNM label again? @rleungx

**Comment by rleungx** [comment]
There is a concurrency problem in the test. @siddontang

**Comment by rleungx** [comment]
The test will be added in another PR. I will create it once this PR is merged.

**Pull request merged**